src/bin/sus-kernel/permission/verify/parsed_sudoers_type.rs

```diff
@@ -0,0 +1,159 @@
+// Since sudoers store users by username and not uid,
+// we use special types in this file to easily query sudoers using ids
+use super::sudoers_type;
+use super::sudoers_type::Command;
+use super::sudoers_type::User::{Useralias, Usergroup, Username};
+use super::Permission;
+use super::ALL;
+use nix::unistd::{Gid, Uid};
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::ffi::CString;
+use users::{get_group_by_name, get_user_by_name};
+
+#[derive(Debug)]
+pub struct AllowedCmd {
+    pub users: HashSet<Uid>,
+    pub allow_all_users: bool,
+    pub groups: HashSet<Gid>,
+    pub allow_all_groups: bool,
+    pub paths: HashSet<CString>,
+    pub allow_all_cmds: bool,
+    pub options: Vec<sudoers_type::Option>,
+}
+
+impl AllowedCmd {
+    pub fn new() -> Self {
+        AllowedCmd {
+            users: HashSet::new(),
+            groups: HashSet::new(),
+            paths: HashSet::new(),
+            options: Vec::new(),
+            allow_all_cmds: false,
+            allow_all_users: false,
+            allow_all_groups: false,
+        }
+    }
+    pub fn is_relevant(&self, req_perm: &Permission) -> bool {
+        self.users.contains(&req_perm.uid)
+            || self.groups.contains(&req_perm.primary_gid)
+            || !self.groups.is_disjoint(&req_perm.secondary_gids)
+            || self.allow_all_users
+            || self.allow_all_groups
+    }
+}
+#[derive(Debug)]
+pub struct Rule {
+    pub users: HashSet<Uid>,
+    pub allow_all_users: bool,
+    pub groups: HashSet<Gid>,
+    pub allow_all_groups: bool,
+    pub allowed_cmds: Vec<AllowedCmd>,
+    pub allow_all_cmds: bool,
+}
+
+fn get_uid_from_username(username: &str) -> Option<Uid> {
+    get_user_by_name(username).map(|user| Uid::from_raw(user.uid()))
+}
+
+fn get_gid_from_groupname(groupname: &str) -> Option<Gid> {
+    get_group_by_name(groupname).map(|user| Gid::from_raw(user.gid()))
+}
+
+impl Rule {
+    pub fn new() -> Self {
+        Rule {
+            users: HashSet::new(),
+            groups: HashSet::new(),
+            allowed_cmds: Vec::new(),
+            allow_all_groups: false,
+            allow_all_users: false,
+            allow_all_cmds: false,
+        }
+    }
```

Comment:
Is this constructor ever used? If not, it should probably be removed.

Reply:
yup, in pub fn from_userspec(), it's used to generate new rules